File: backend/app/schemas/invitation.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

# --- Event Schemas ---
class EventBase(BaseModel):
    title: str
    date: str
    time: str
    venue: str
    description: Optional[str] = None
    display_order: int = 0

class EventCreate(EventBase):
    pass

class EventUpdate(EventBase):
    pass

class EventOut(EventBase):
    id: str
    invitation_id: str

    class Config:
        from_attributes = True

# --- Family Schemas ---
class FamilyMemberBase(BaseModel):
    name: str
    relation: str
    side: str = "both"  # "bride", "groom", "both"
    display_order: int = 0

class FamilyMemberCreate(FamilyMemberBase):
    pass

class FamilyMemberUpdate(FamilyMemberBase):
    pass

class FamilyMemberOut(FamilyMemberBase):
    id: str
    invitation_id: str

    class Config:
        from_attributes = True

# --- Gallery Schemas ---
class GalleryImageBase(BaseModel):
    image_url: str
    display_order: int = 0

class GalleryImageCreate(GalleryImageBase):
    pass

class GalleryImageOut(GalleryImageBase):
    id: str
    invitation_id: str

    class Config:
        from_attributes = True

# --- Invitation Schemas ---
class InvitationBase(BaseModel):
    groom_name: str
    bride_name: str
    groom_photo: Optional[str] = None
    bride_photo: Optional[str] = None
    couple_photo: Optional[str] = None
    wedding_date: str
    wedding_time: str
    venue_name: str
    venue_address: str
    google_maps_url: Optional[str] = None
    template_id: str = "traditional"
    background_music_url: Optional[str] = None
    invitation_text: Optional[str] = None
    image_transforms: Optional[str] = None
    status: str = "DRAFT"
    slug: Optional[str] = None

class InvitationCreate(InvitationBase):
    events: Optional[List[EventCreate]] = []
    family_members: Optional[List[FamilyMemberCreate]] = []
    gallery_images: Optional[List[GalleryImageCreate]] = []

class InvitationUpdate(InvitationBase):
    events: Optional[List[EventCreate]] = None
    family_members: Optional[List[FamilyMemberCreate]] = None
    gallery_images: Optional[List[GalleryImageCreate]] = None

class InvitationOut(InvitationBase):
    id: str
    slug: str
    created_at: datetime
    updated_at: datetime
    events: List[EventOut] = []
    family_members: List[FamilyMemberOut] = []
    gallery_images: List[GalleryImageOut] = []

    class Config:
        from_attributes = True

class DashboardStats(BaseModel):
    total_invitations: int
    draft_invitations: int
    published_invitations: int

File: src/types/invitation.ts
export interface EventItem {
  id?: string;
  title: string;
  date: string;
  time: string;
  venue: string;
  description?: string;
  display_order?: number;
}

export interface FamilyMemberItem {
  id?: string;
  name: string;
  relation: string;
  side: "bride" | "groom" | "both";
  display_order?: number;
}

export interface GalleryImageItem {
  id?: string;
  image_url: string;
  display_order?: number;
}

export type TemplateId =
  | "traditional"
  | "royal"
  | "floral"
  | "kalamkari"
  | "banarasi"
  | "gold-red"
  | "green-gold"
  | "vintage"
  | "pattachitra"
  | "luxury"
  | "pastel"
  | "contemporary"
  | "cinematic"
  | "family"
  | "festive"
  | "engagement"
  | "reception"
  | "haldi"
  | "sangeet"
  | "sacred-vivaha";

export interface ImageTransform {
  x?: number;        // Position X in px relative to element default
  y?: number;        // Position Y in px relative to element default
  scale?: number;    // Scale multiplier (1.0 = 100%)
  rotation?: number; // Rotation in degrees (-180 to +180)
  opacity?: number;  // Opacity (0.0 to 1.0)
  zIndex?: number;   // Layer ordering
  cropX?: number;    // Object offset X (%)
  cropY?: number;    // Object offset Y (%)
}

export interface InvitationData {
  id?: string;
  slug?: string;
  groom_name: string;
  bride_name: string;
  groom_photo?: string;
  bride_photo?: string;
  couple_photo?: string;
  wedding_date: string;
  wedding_time: string;
  venue_name: string;
  venue_address: string;
  google_maps_url?: string;
  template_id: TemplateId;
  background_music_url?: string;
  invitation_text?: string;
  image_transforms?: Record<string, ImageTransform> | string;
  status: "DRAFT" | "PUBLISHED";
  created_at?: string;
  updated_at?: string;
  events: EventItem[];
  family_members: FamilyMemberItem[];
  gallery_images: GalleryImageItem[];
}

export interface DashboardStats {
  total_invitations: number;
  draft_invitations: number;
  published_invitations: number;
}
